src/raftstore/coprocessor/region_snapshot.rs

```diff
@@ -131,13 +96,122 @@ impl<'a> Peekable for RegionSnapshot<'a> {
     }
 }
 
+/// `RegionIterator` wrap a rocksdb iterator and only allow it to
+/// iterate in the region. It behaves as if underlying
+/// db only contains one region.
+pub struct RegionIterator<'a> {
+    iter: DBIterator<'a>,
+    valid: bool,
+    region: Region,
+    start_key: Vec<u8>,
+    end_key: Vec<u8>,
+}
+
+// we use rocksdb's style iterator, so omit the warning.
+#[allow(should_implement_trait)]
+impl<'a> RegionIterator<'a> {
+    pub fn new(iter: DBIterator<'a>, region: Region) -> RegionIterator<'a> {
+        RegionIterator {
+            iter: iter,
+            valid: false,
+            start_key: keys::enc_start_key(&region),
+            end_key: keys::enc_end_key(&region),
+            region: region,
+        }
+    }
+
+    pub fn seek_to_first(&mut self) -> bool {
+        self.valid = self.iter.seek(self.start_key.as_slice().into());
+
+        self.update_valid()
+    }
+
+    #[inline]
+    fn update_valid(&mut self) -> bool {
+        if self.valid {
+            let key = self.iter.key();
+            self.valid = key >= &self.start_key && key < &self.end_key;
+        }
+        self.valid
+    }
+
+    pub fn seek_to_last(&mut self) -> bool {
+        if !self.iter.seek(self.end_key.as_slice().into()) && !self.iter.seek(SeekKey::End) {
+            self.valid = false;
+            return self.valid;
+        }
+
+        while self.iter.key() >= &self.end_key && self.iter.prev() {
+        }
+
+        self.valid = self.iter.valid();
+        self.update_valid()
+    }
+
+    pub fn seek(&mut self, key: &[u8]) -> Result<bool> {
+        let key = try!(self.should_seekable(key));
+        if key == self.end_key {
+            self.valid = false;
+        } else {
+            self.valid = self.iter.seek(key.as_slice().into());
+        }
+
+        Ok(self.update_valid())
+    }
+
+    pub fn prev(&mut self) -> bool {
+        if !self.valid {
+            return false;
+        }
+        self.valid = self.iter.prev();
+
+        self.update_valid()
+    }
+
+    pub fn next(&mut self) -> bool {
+        if !self.valid {
+            return false;
+        }
+        self.valid = self.iter.next();
+
+        self.update_valid()
+    }
+
+    #[inline]
+    pub fn key(&self) -> &[u8] {
+        assert!(self.valid);
+        keys::origin_key(self.iter.key())
+    }
+
+    #[inline]
+    pub fn value(&self) -> &[u8] {
+        assert!(self.valid);
+        self.iter.value()
+    }
+
+    #[inline]
+    pub fn valid(&self) -> bool {
+        self.valid
+    }
+
+    #[inline]
+    fn should_seekable(&self, key: &[u8]) -> Result<Vec<u8>> {
+        let key = keys::data_key(key);
+        if key < self.start_key || key > self.end_key {
```

Comment:
why key == end_key is not KeyNotInRegion error?
The region range is [start, end), so I think end is also KeyNotInRegion.